go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,418 @@
+	klog.Infof("Starting cluster manager")
+	metrics.RegisterClusterManagerFunctional()
+
+	existingNodes, err := cm.kube.GetNodes()
+	if err != nil {
+		klog.Errorf("Error in fetching nodes: %v", err)
+		return err
+	}
+	klog.V(5).Infof("Existing number of nodes: %d", len(existingNodes.Items))
+
+	klog.Infof("Allocating subnets")
+	if err := cm.InitSubnetAllocatorRanges(config.Default.ClusterSubnets); err != nil {
+		klog.Errorf("Failed to initialize host subnet allocator ranges: %v", err)
+		return err
+	}
+
+	if config.HybridOverlay.Enabled {
+		if err := cm.InitHybridOverlaySubnetAllocatorRanges(config.HybridOverlay.ClusterSubnets); err != nil {
+			klog.Errorf("Failed to initialize hybrid overlay subnet allocator ranges: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cm *ClusterManager) InitSubnetAllocatorRanges(subnets []config.CIDRNetworkEntry) error {
+	return cm.clusterSubnetAllocator.InitRanges(subnets)
+}
+
+func (cm *ClusterManager) InitHybridOverlaySubnetAllocatorRanges(subnets []config.CIDRNetworkEntry) error {
+	return cm.hybridOverlaySubnetAllocator.InitRanges(subnets)
+}
+
+// Run starts the actual watching.
+func (cm *ClusterManager) Run(nodeName string) error {
+	// Start and sync the watch factory to begin listening for events
+	if err := cm.watchFactory.Start(); err != nil {
+		return err
+	}
+
+	if err := cm.WatchNodes(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// WatchNodes starts the watching of node resource and calls
+// back the appropriate handler logic
+func (cm *ClusterManager) WatchNodes() error {
+	_, err := cm.retryNodes.WatchResource()
+	return err
+}
+
+func (cm *ClusterManager) syncNodes(nodes []interface{}) error {
+	for _, tmp := range nodes {
+		node, ok := tmp.(*kapi.Node)
+		if !ok {
+			return fmt.Errorf("spurious object in syncNodes: %v", tmp)
+		}
+
+		hostSubnets, _ := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+		if config.HybridOverlay.Enabled && len(hostSubnets) == 0 && houtil.IsHybridOverlayNode(node) {
+			// this is a hybrid overlay node so mark as allocated from the hybrid overlay subnet allocator
+			hostSubnet, err := houtil.ParseHybridOverlayHostSubnet(node)
+			if err != nil {
+				klog.Warning(err.Error())
+			} else if hostSubnet != nil {
+				klog.V(5).Infof("Node %s contains subnets: %v", node.Name, hostSubnet)
+				if err := cm.hybridOverlaySubnetAllocator.MarkSubnetsAllocated(node.Name, hostSubnet); err != nil {
+					utilruntime.HandleError(err)
+				}
+			}
+			// there is nothing left to be done if this is a hybrid overlay node
+			continue
+		}
+
+		klog.V(5).Infof("Node %s contains subnets: %v", node.Name, hostSubnets)
+		if err := cm.clusterSubnetAllocator.MarkSubnetsAllocated(node.Name, hostSubnets...); err != nil {
+			utilruntime.HandleError(err)
+		}
+	}
+	return nil
+}
+
+func (cm *ClusterManager) addUpdateNodeEvent(node *kapi.Node) error {
+	if noHostSubnet := util.NoHostSubnet(node); noHostSubnet {
```

Comment:
here we can shorten it with just: `if util.NoHostSubnet(node) {`

Reply:
Ack